sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala

```diff
@@ -566,7 +566,7 @@ class DateFunctionsSuite extends QueryTest with SharedSQLContext {
 
     // february
     val y1 = "2016-02-29"
-    val y2 = "2017-02-29"
+    val y2 = "2017-02-32"
```

Comment:
Do you roughly know why this has to be changed? Seems like `2017-02-29` date isn't existent.

Reply:
By default (in [SMART](https://docs.oracle.com/javase/8/docs/api/java/time/format/ResolverStyle.html#SMART) resolving style), new parser just checks that values fall to valid ranges. So, it is 1-31 for `dd`.  The parsed date is "casted" to closer valid date later. In this particular case, `2017-02-29` is interpreted as a valid date because `29` is in the range of `[1,31]`, and it is converted to number of microseconds/days since epoch for `2017-02-28`.

Reply:
Hm .. sounds it's too much SMART :)

 Would it be better to set the mode to STRICT then? I think we should set the mode that breaks as less as possible.

Reply:
adding @cloud-fan as well

Reply:
> Would it be better to set the mode to STRICT then?

I don't know how strict the `STRICT` mode. If I change it now in the PR, most likely I have to made unrelated to this PR modifications. I would prefer to do that in a separate PR.

Reply:
Hm, okie. Would you be interested in giving a try to change mode in a separate PR?

Reply:
> Would you be interested in giving a try to change mode in a separate PR?

Yes, I am interested in it. I will try to change it on the current branch to see how much it is strict.

Reply:
I enabled the STRICT mode in the e4baad6e5a445bf85c312e606fbc8d55eaf0eb37 commit. Let's see how it will pass all tests.

Reply:
@HyukjinKwon Finally all test on the strict mode passed. Please, take a look at the PR one more time.